File: Recurrsion/problem5.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/*
 Problem:
 Given an integer array of size N, find all occurrences (indices) of a given element (key)
 and print them. Solve this using a recursive function.

 This file: problem5.java
 Usage (stdin):
    - First integer: N (size of array)
    - Next N integers: array elements
    - Last integer: key to search for
 Example:
    Input:
        6
        1 2 3 2 4 2
        2
    Output:
        1 3 5
*/

public class problem5{
    public static void FindAllOccurrences(int [] arr, int key, int idx, List<Integer> occurrences){
        if(idx==arr.length){
            return;
        }
        if(arr[idx]==key){
            occurrences.add(idx);
        }
        FindAllOccurrences(arr, key, idx + 1, occurrences);
        
    }
    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        int n=sc.nextInt();
        int [] arr=new int[n];
        for(int i=0;i<n;i++){
            arr[i]=sc.nextInt();
        }
        int key=sc.nextInt();
        List<Integer> occurrences=new ArrayList<>();
        FindAllOccurrences(arr, key, 0, occurrences);
        for(int index: occurrences){
            System.out.print(index+" ");
        }
        sc.close();
    }
}

File: CPP(STL)/problem-5.cpp
/*

4️⃣ Find Element

Concepts:

find()

iterator

Problem

Return index of x in array. If not present print -1.

Example

Input
5
10 20 30 40 50
30

Output

2

Hint

auto it = find(v.begin(), v.end(), x);


*/
#include<iostream>
#include<algorithm>
#include<vector>
using namespace std;
int main(){
    int n;
    cout<<"Enter the size of the array: ";
    cin>>n;
    vector<int> arr(n);
    cout<<"Enter the elements of the array: ";
    for(int i=0; i<n; i++){
        cin>>arr[i];
    }
    int x;
    cout<<"Enter the number to find its index: ";
    cin>>x;
    
    auto it = find(arr.begin(), arr.end(), x);
    
    if(it != arr.end()){
        int index = distance(arr.begin(), it);
        cout<<"The index of "<<x<<" is: "<<index<<endl;
    } else {
        cout<<"-1"<<endl; // Element not found
    }
}